core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -288,19 +294,24 @@ private[spark] class Executor(
     threadPool.shutdown()
 
     // Notify plugins that executor is shutting down so they can terminate cleanly
+    pluginShutdown
+    if (!isLocal) {
+      env.stop()
+    }
+  }
+
+  private def pluginShutdown: Unit = {
     Utils.withContextClassLoader(replClassLoader) {
       executorPlugins.foreach { plugin =>
         try {
+          executorShutdown = true
```

Comment:
I think you need to make this volatile, and check it at the start of the method, and only execute if false, and set it to true before starting the other work.